File: Hard/01FourNumberSum.py
def fourNumberSum(array, targetSum):
    allPairSum = {}
    quadruplets = []
    for i in range(1, len(array) - 1):
        for j in range(i+1, len(array)):
            currentSum = array[i] + array[j]
            difference = targetSum - currentSum
            if difference in allPairSum:
                for pair in allPairSum[difference]:
                    quadruplets.append(pair + [array[i], array[j]])

        for k in range(0, i):
            currentSum = array[i] + array[k]
            if currentSum not in allPairSum:
                allPairSum[currentSum] = [[array[k], array[i]]]
            else:
                allPairSum[currentSum].append([array[k], array[i]])

    return quadruplets

File: Hard/01FourNumberSum.js
function fourNumberSum(array, targetSum) {
  const allPairSum = {};
  const quadruplets = [];

  for (let i = 1; i < array.length - 1; i++) {
    for (let j = i + 1; j < array.length; j++) {
      const currentSum = array[i] + array[j];
      const difference = targetSum - currentSum;

      if (difference in allPairSum) {
        for (const pair of allPairSum[difference]) {
          quadruplets.push([...pair, array[i], array[j]]);
        }
      }
    }

    for (let k = 0; k < i; k++) {
      const currentSum = array[i] + array[k];

      if (!allPairSum[currentSum]) {
        allPairSum[currentSum] = [[array[k], array[i]]];
      } else {
        allPairSum[currentSum].push([array[k], array[i]]);
      }
    }
  }

  return quadruplets;
}
